src/core/featurehistory.h

```diff
@@ -0,0 +1,124 @@
+    Q_PROPERTY( bool isRedoEnabled READ isRedoEnabled NOTIFY isRedoEnabledChanged )
+
+  public:
+    /**
+     * Stores the created, updated and deleted features on each undo/redo step.
+     */
+    struct FeatureModifications
+    {
+        FeatureModifications()
+        {}
+
+        QList<OldNewFeaturePair> createdFeatures;
+        QList<OldNewFeaturePair> updatedFeatures;
+        QList<OldNewFeaturePair> deletedFeatures;
+    };
+
+    /**
+     * Construct a new Feature history object
+     *
+     * @param project
+     */
+    explicit FeatureHistory( const QgsProject *project );
+
+    //! Perform undo of the most recent modification step
+    Q_INVOKABLE bool undo();
+
+    //! Perform redo of the most recent modification step
+    Q_INVOKABLE bool redo();
+
+    //! Get the undo message to be show in the UI. NOTE should be called before calling \a undo.
+    Q_INVOKABLE const QString undoMessage();
+
+    //! Get the redo message to be show in the UI. NOTE should be called before calling \a redo.
+    Q_INVOKABLE const QString redoMessage();
+
+    bool isUndoEnabled();
+    bool isRedoEnabled();
+
+  private:
+    static const int sTimeoutMs = 50;
+
+    //! Add the needed event listeners to monitor for changes.
+    void addLayerListeners();
+
+    //! Apply given modifications on all layers in the current project. Used both by undo and redo operations.
+    bool applyModifications( QMap<QString, FeatureModifications> &modificationsByLayerId );
+
+    //! Reverse the modification. Used to make undo modifications into redo modifications.
+    QMap<QString, FeatureModifications> reverseModifications( QMap<QString, FeatureModifications> &modificationsByLayerId );
+
+    //! The current project instance.
+    const QgsProject *mProject = nullptr;
+
+    //! If currently applying undo or redo feature modifications.
+    bool mIsApplyingModifications = false;
+
+    //! Timer to wait for short time before all features (and child features) are saved. When timeouts, all temporary modifications are added as a new undo step.
+    QTimer mTimer;
+
+    //! Temporary storage of all modifications before creating a new undo step.
+    QMap<QString, FeatureModifications> mTempHistoryStep;
+
+    //! Temporary storage of All features that have been modified before creating a new undo step.
+    QMap<QString, QMap<QgsFeatureId, QgsFeature>> mTempModifiedFeaturesByLayerId;
+
+    //! Temporary storage of the deleted feature ids before creating a new undo step.
+    QMap<QString, QgsFeatureIds> mTempDeletedFeatureIdsByLayerId;
+
+    //! Undo history records
+    QList<QMap<QString, FeatureModifications>> mUndoHistory;
+
+    //! Redo history records
+    QList<QMap<QString, FeatureModifications>> mRedoHistory;
+
+    //! Layer ids being observed for changes. Should reset when the project is changed. Used to prevent double event listeners.
+    QSet<QString> mObservedLayerIds;
+
+  signals:
```

Comment:
Throughout QGIS and our code, we put signals and private slots _above private members.